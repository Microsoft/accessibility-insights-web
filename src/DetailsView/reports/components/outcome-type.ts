// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { countBy, values } from 'lodash';

import { ManualTestStatus, ManualTestStatusData } from '../../../common/types/manual-test-status';

export type OutcomeType = 'pass' | 'incomplete' | 'fail';

export const allOutcomeTypes: OutcomeType[] = ['pass', 'incomplete', 'fail'];

export function outcomeTypeFromTestStatus(testStatus: ManualTestStatus) {
    const statusMap = {
        [ManualTestStatus.PASS]: 'pass',
        [ManualTestStatus.FAIL]: 'fail',
        [ManualTestStatus.UNKNOWN]: 'incomplete',
    } as { [P in ManualTestStatus]: OutcomeType };

    return statusMap[testStatus];
}

export function outcomeTypeSemanticsFromTestStatus(testStatus: ManualTestStatus): OutcomeTypeSemantic {
    return outcomeTypeSemantics[outcomeTypeFromTestStatus(testStatus)];
}

export interface OutcomeTypeSemantic {
    pastTense: string;
}

export const outcomeTypeSemantics: { [OT in OutcomeType]: OutcomeTypeSemantic } = {
    pass: { pastTense: 'passed' },
    incomplete: { pastTense: 'incomplete' },
    fail: { pastTense: 'failed' },
};

export type OutcomeStats = { [OT in OutcomeType]: number };

export function outcomeStatsFromManualTestStatus(testStepStatus: ManualTestStatusData): OutcomeStats {
    const outcomeTypeSet = values(testStepStatus).map(s => outcomeTypeFromTestStatus(s.stepFinalResult));
    const stats = countBy(outcomeTypeSet) as OutcomeStats;
    stats.pass = stats.pass || 0;
    stats.incomplete = stats.incomplete || 0;
    stats.fail = stats.fail || 0;
    return stats;
}
