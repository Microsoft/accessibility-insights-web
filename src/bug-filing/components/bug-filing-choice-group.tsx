// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ChoiceGroup, IChoiceGroupOption } from 'office-ui-fabric-react/lib/ChoiceGroup';
import * as React from 'react';

import { UserConfigMessageCreator } from '../../common/message-creators/user-config-message-creator';
import { NamedSFC } from '../../common/react/named-sfc';
import { BugFilingService } from '../types/bug-filing-service';

export type BugFilingChoiceGroupProps = {
    deps: BugFilingChoiceGroupDeps;
    selectedBugFilingService: BugFilingService;
    bugFilingServices: BugFilingService[];
};

export type BugFilingChoiceGroupDeps = {
    userConfigMessageCreator: UserConfigMessageCreator;
};

export const BugFilingChoiceGroup = NamedSFC<BugFilingChoiceGroupProps>('BugFilingChoiceGroup', props => {
    const { userConfigMessageCreator } = props.deps;

    const getOptions: () => IChoiceGroupOption[] = () => {
        return props.bugFilingServices.map(service => {
            return {
                key: service.key,
                text: service.displayName,
            };
        });
    };

    const onChange = (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, option?: IChoiceGroupOption) => {
        userConfigMessageCreator.setBugService(option.key);
    };

    return (
        <ChoiceGroup
            className={'bug-filing-choice-group'}
            onChange={onChange}
            options={getOptions()}
            selectedKey={props.selectedBugFilingService.key}
        />
    );
});
