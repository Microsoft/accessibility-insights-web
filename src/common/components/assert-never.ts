// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// This function helps with exhaustiveness checking in switch statements
// If used in the default case, TS will throw an error unless all cases are handled.
export const assertNever = (x: never): never => {
    throw Error(
        'This error should never be thrown when strict null checks are applied to this project',
    );
};
