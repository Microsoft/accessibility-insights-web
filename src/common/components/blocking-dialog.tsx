// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DefaultButton, PrimaryButton } from 'office-ui-fabric-react/lib/Button';
import { Dialog, DialogFooter, DialogType, IDialogProps } from 'office-ui-fabric-react/lib/Dialog';
import * as React from 'react';

import { NamedSFC } from '../../common/react/named-sfc';

export type BlockingDialogProps = IDialogProps;

export const BlockingDialog = NamedSFC<BlockingDialogProps>('BlockingDialog', props => {
    return (
        // As of writing, NVDA has an issue where it doesn't announce new elements with role=alertdialog,
        // which Office Fabric applies implicitly whenever isBlocking: true is set. As a workaround, our
        // "blocking" dialog sets isBlocking: false (which means the dialog will use role=dialog instead of
        // role=alertdialog, which NVDA announces more reliably), and emulates being "blocking" by making
        // sure we don't set any onDismiss handlers and overriding showCloseButton to false.
        <Dialog
            {...props}
            onDismiss={undefined}
            dialogContentProps={{
                ...props.dialogContentProps,
                showCloseButton: false,
            }}
            modalProps={{
                ...props.modalProps,
                // Be warned that changing this value can change which child element is focused by default
                // (we saw this in the telemetry-permission-dialog); be sure to test this behavior specifically
                // if you ever change this back to true.
                isBlocking: false,
                onDismiss: undefined,
            }}
        />
    );
});
