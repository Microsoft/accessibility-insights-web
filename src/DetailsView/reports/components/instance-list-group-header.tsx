// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { GuidanceLinks } from '../../../common/components/guidance-links';
import { NewTabLink } from '../../../common/components/new-tab-link';
import { NamedSFC } from '../../../common/react/named-sfc';
import { RuleResult } from '../../../scanner/iruleresults';
import { OutcomeChip } from './outcome-chip';
import { OutcomeType } from './outcome-type';

export interface InstaceListGroupHeaderProps {
    ruleResult: RuleResult;
    outcomeType: OutcomeType;
}

export const InstanceListGroupHeader = NamedSFC<InstaceListGroupHeaderProps>('InstaceListGroupHeader', props => {
    const renderCountBadge = () => {
        const { outcomeType, ruleResult } = props;

        if (outcomeType !== 'fail') {
            return null;
        }

        return <OutcomeChip count={ruleResult.nodes.length} outcomeType={outcomeType} />;
    };

    const renderRuleLink = () => {
        const ruleResult = props.ruleResult;
        const ruleId = ruleResult.id;
        const ruleUrl = ruleResult.helpUrl;
        return (
            <span className="rule-details-id">
                <NewTabLink href={ruleUrl} aria-label={`rule ${ruleId}`} aria-describedby={`${ruleId}-rule-description`}>
                    {ruleId}
                </NewTabLink>
            </span>
        );
    };

    const renderGuidanceLinks = () => {
        return <GuidanceLinks links={props.ruleResult.guidanceLinks} />;
    };

    const renderDescription = () => {
        return <span className="rule-details-description">{props.ruleResult.description}</span>;
    };

    return (
        <div role="heading">
            {renderCountBadge()} {renderRuleLink()}: {renderDescription()} ({renderGuidanceLinks()})
        </div>
    );
});
