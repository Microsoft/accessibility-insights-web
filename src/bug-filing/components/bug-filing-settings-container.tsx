// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedSFC } from '../../common/react/named-sfc';
import { BugServiceProperties } from '../../common/types/store-data/user-configuration-store';
import { SettingsDeps } from '../../DetailsView/components/settings-panel/settings/settings-props';
import { BugFilingServiceProvider } from '../bug-filing-service-provider';
import { BugFilingService } from '../types/bug-filing-service';
import { BugFilingChoiceGroup, BugFilingChoiceGroupDeps } from './bug-filing-choice-group';

export interface BugFilingSettingsContainerProps {
    deps: BugFilingSettingsContainerDeps;
    selectedBugFilingService: BugFilingService;
    selectedBugFilingServiceData: BugServiceProperties;
}

export type BugFilingSettingsContainerDeps = {
    bugServiceProvider: BugFilingServiceProvider;
} & BugFilingChoiceGroupDeps &
    SettingsDeps;

export const BugFilingSettingsContainer = NamedSFC<BugFilingSettingsContainerProps>('BugFilingSettingsContainer', props => {
    const { deps, selectedBugFilingService, selectedBugFilingServiceData } = props;
    const SettingsForm = selectedBugFilingService.renderSettingsForm;
    const bugFilingServices = deps.bugServiceProvider.allVisible();

    return (
        <>
            <BugFilingChoiceGroup deps={deps} selectedBugFilingService={selectedBugFilingService} bugFilingServices={bugFilingServices} />
            <SettingsForm deps={deps} settings={selectedBugFilingServiceData} />
        </>
    );
});
