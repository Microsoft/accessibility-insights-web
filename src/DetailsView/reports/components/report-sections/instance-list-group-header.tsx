// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { kebabCase } from 'lodash';
import * as React from 'react';

import { GuidanceLinks } from '../../../../common/components/guidance-links';
import { GuidanceTags } from '../../../../common/components/guidance-tags';
import { NewTabLink } from '../../../../common/components/new-tab-link';
import { GetGuidanceTagsFromGuidanceLinks } from '../../../../common/get-guidance-tags-from-guidance-links';
import { NamedSFC } from '../../../../common/react/named-sfc';
import { RuleResult } from '../../../../scanner/iruleresults';
import { InstanceOutcomeType } from '../instance-outcome-type';
import { OutcomeChip } from '../outcome-chip';
import { outcomeTypeSemantics } from '../outcome-type';

export type InstanceListGroupHeaderDeps = {
    getGuidanceTagsFromGuidanceLinks: GetGuidanceTagsFromGuidanceLinks;
};

export interface InstanceListGroupHeaderProps {
    deps: InstanceListGroupHeaderDeps;
    ruleResult: RuleResult;
    outcomeType: InstanceOutcomeType;
    ariaLevel?: number;
}

export const InstanceListGroupHeader = NamedSFC<InstanceListGroupHeaderProps>('InstaceListGroupHeader', props => {
    const outcomeText = outcomeTypeSemantics[props.outcomeType].pastTense;
    const ariaDescribedBy = `${kebabCase(outcomeText)}-rule-${props.ruleResult.id}-description`;

    const renderCountBadge = () => {
        const { outcomeType, ruleResult } = props;

        if (outcomeType !== 'fail') {
            return null;
        }

        return <OutcomeChip count={ruleResult.nodes.length} outcomeType={outcomeType} />;
    };

    const renderRuleLink = () => {
        const ruleResult = props.ruleResult;
        const ruleId = ruleResult.id;
        const ruleUrl = ruleResult.helpUrl;
        return (
            <span className="rule-details-id">
                <NewTabLink href={ruleUrl} aria-label={`rule ${ruleId}`} aria-describedby={ariaDescribedBy}>
                    {ruleId}
                </NewTabLink>
            </span>
        );
    };

    const renderGuidanceLinks = () => {
        return <GuidanceLinks links={props.ruleResult.guidanceLinks} />;
    };

    const renderDescription = () => {
        return (
            <span className="rule-details-description" id={ariaDescribedBy}>
                {props.ruleResult.description}
            </span>
        );
    };

    const renderGuidanceTags = () => {
        return <GuidanceTags deps={props.deps} links={props.ruleResult.guidanceLinks} />;
    };

    const headingProps =
        props.ariaLevel != null
            ? {
                  role: 'heading',
                  'aria-level': props.ariaLevel,
              }
            : null;

    return (
        <div {...headingProps}>
            {renderCountBadge()} {renderRuleLink()}: {renderDescription()} ({renderGuidanceLinks()}) {renderGuidanceTags()}
        </div>
    );
});
