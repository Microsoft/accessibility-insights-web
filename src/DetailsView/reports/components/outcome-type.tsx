// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import { InapplicableIcon } from '../../../common/icons/inapplicable-icon';
import { CheckIcon } from './../../../common/icons/check-icon';
import { CircleIcon } from './../../../common/icons/circle-icon';
import { CrossIcon } from './../../../common/icons/cross-icon';
import { InstanceOutcomeType } from './report-sections/outcome-summary-bar';
import { RequirementOutcomeType } from './requirement-outcome-type';

export type OutcomeType = RequirementOutcomeType | InstanceOutcomeType;

export interface OutcomeTypeSemantic {
    pastTense: string;
}

export const outcomeTypeSemantics: { [OT in OutcomeType]: OutcomeTypeSemantic } = {
    pass: { pastTense: 'Passed' },
    incomplete: { pastTense: 'Incomplete' },
    fail: { pastTense: 'Failed' },
    inapplicable: { pastTense: 'Not applicable' },
};

export const outcomeIconMap: { [OT in OutcomeType]: JSX.Element } = {
    pass: <CheckIcon />,
    incomplete: <CircleIcon />,
    fail: <CrossIcon />,
    inapplicable: <InapplicableIcon />,
};
