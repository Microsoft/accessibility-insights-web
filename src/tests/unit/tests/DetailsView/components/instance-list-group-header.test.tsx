// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import * as React from 'react';

import { InstanceListGroupHeader } from '../../../../../DetailsView/reports/components/instance-list-group-header';
import { OutcomeType } from '../../../../../DetailsView/reports/components/outcome-type';
import { RuleResult } from '../../../../../scanner/iruleresults';

describe('InstanceListGroupHeaderTest', () => {
    const getSampleRuleResult = () => {
        const ruleResult: RuleResult = {
            id: 'ruleid',
            nodes: [{} as AxeNodeResult],
            description: 'rule description',
            helpUrl: 'help url',
        };
        return ruleResult;
    };

    test.each(['pass', 'fail'])('render for %s instances', (outcome: OutcomeType) => {
        const wrapper = shallow(<InstanceListGroupHeader ruleResult={getSampleRuleResult()} outcome={outcome} />);
        expect(wrapper.getElement()).toMatchSnapshot();
    });
});
