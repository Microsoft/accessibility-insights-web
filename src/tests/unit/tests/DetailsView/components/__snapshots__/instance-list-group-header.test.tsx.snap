// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstanceListGroupHeaderTest render for fail instances 1`] = `
<div
  role="heading"
>
  <OutcomeChip
    count={1}
    outcomeType="fail"
  />
   
  <span
    className="rule-details-id"
  >
    <NewTabLink
      aria-describedby="ruleid-rule-description"
      aria-label="rule ruleid"
      href="help url"
    >
      ruleid
    </NewTabLink>
  </span>
  : 
  <span
    className="rule-details-description"
  >
    rule description
  </span>
   (
  <GuidanceLinks />
  )
</div>
`;

exports[`InstanceListGroupHeaderTest render for incomplete instances 1`] = `
<div
  role="heading"
>
   
  <span
    className="rule-details-id"
  >
    <NewTabLink
      aria-describedby="ruleid-rule-description"
      aria-label="rule ruleid"
      href="help url"
    >
      ruleid
    </NewTabLink>
  </span>
  : 
  <span
    className="rule-details-description"
  >
    rule description
  </span>
   (
  <GuidanceLinks />
  )
</div>
`;

exports[`InstanceListGroupHeaderTest render for pass instances 1`] = `
<div
  role="heading"
>
   
  <span
    className="rule-details-id"
  >
    <NewTabLink
      aria-describedby="ruleid-rule-description"
      aria-label="rule ruleid"
      href="help url"
    >
      ruleid
    </NewTabLink>
  </span>
  : 
  <span
    className="rule-details-description"
  >
    rule description
  </span>
   (
  <GuidanceLinks />
  )
</div>
`;
